drivers/gpio/gpio_silabs_siwx917.c

```diff
@@ -0,0 +1,414 @@
+		}
+	}
+	return status;
+}
+
+static const struct gpio_driver_api gpio_siwx917_driver_api = {
+	.pin_configure = gpio_siwx917_pin_configure,
+	.port_get_raw = gpio_siwx917_port_get,
+	.port_set_masked_raw = gpio_siwx917_port_set_masked,
+	.port_set_bits_raw = gpio_siwx917_port_set_bits,
+	.port_clear_bits_raw = gpio_siwx917_port_clear_bits,
+	.port_toggle_bits = gpio_siwx917_port_toggle_bits,
+	.pin_interrupt_configure = gpio_siwx917_interrupt_configure,
+	.manage_callback = gpio_siwx917_manage_callback,
+	.get_pending_int = gpio_siwx917_get_pending_int,
+#ifdef CONFIG_GPIO_GET_DIRECTION
+	.port_get_direction = gpio_siwx917_port_get_direction,
+#endif
+};
+
+#define GPIO_PORT_INIT(n)                                                                          \
+	static int gpio_siwx917_init_port_##n##_(const struct device *dev);                        \
+	static const struct gpio_siwx917_port_config gpio_siwx917_port_##n##_config = {            \
+		.common =                                                                          \
+			{                                                                          \
+				.port_pin_mask = GPIO_PORT_PIN_MASK_FROM_DT_NODE(n),               \
+			},                                                                         \
+		.parent = DEVICE_DT_GET(DT_PARENT(n)),                                             \
+		.pads = DT_PROP(n, silabs_pads),                                                   \
+		.port = DT_REG_ADDR(n),                                                            \
+		.hal_port = (DT_PROP(DT_PARENT(n), silabs_ulp) ? SL_GPIO_ULP_PORT : 0) +           \
+			    DT_REG_ADDR(n),                                                        \
+		.ulp = DT_PROP(DT_PARENT(n), silabs_ulp),                                          \
+	};                                                                                         \
+	static struct gpio_siwx917_port_data gpio_siwx917_port_##n##_data;                         \
+                                                                                                   \
+	DEVICE_DT_DEFINE(n, gpio_siwx917_init_port_##n##_, NULL, &gpio_siwx917_port_##n##_data,    \
+			 &gpio_siwx917_port_##n##_config, PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY, \
+			 &gpio_siwx917_driver_api);                                                \
+	static int gpio_siwx917_init_port_##n##_(const struct device *dev)                         \
+	{                                                                                          \
+		const struct gpio_siwx917_port_config *config = dev->config;                       \
+		const struct device *parent = config->parent;                                      \
+		struct gpio_siwx917_common_data *data = parent->data;                              \
+		gpio_siwx917_add_port(data, dev);                                                  \
+		return 0;                                                                          \
+	}
```

Comment:
It seems this function does not need to be templated. With a small change in the prototype of `gpio_siwx917_add_port()`, probably we could get rid of `gpio_siwx917_init_port_xxx()`.

Reply:
Good point, simplified.